### PR TITLE
Fix: Avoid magic numbers in rule options (fixes #4182)

### LINKED ISSUE #4182: Avoid magic numbers in rule options
if 
 - the configuration files are js files
 - "no-magic-numbers" is set to 2
 - ESlint version is 1.7.1

then the configuration of some rules flag an error (_"No magic number: …"_). So far, in my case:
 - complexity
 - max-depth
 - max-len
 - max-params
 - max-statements
 - max-nested-callbacks

The reason is these rules expect numbers in their configuration, e.g.
 `"max-nested-callbacks" : [2, 3]`
"3" is the problem, "2" is not since it's in the default list of the "no-magic-numbers" rule.

This is surely nitpicking, but wouldn't it be the occasion to have a proper variable name, eg
`"max-nested-callbacks" : [2, {“maximum" : 3}]`
`"complexity"  : [2, {"maximum" : 5}]`
etc ?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27499861-avoid-magic-numbers-in-rule-options?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by bsbeeks**

### CODE REVIEW

**Comment by ilyavolodin** [comment]
Thanks! LGTM

**Comment by bsbeeks** [comment]
Yeah, I was trying to figure out why it was failing and couldn't see anything in the logs. I'll look again..but maybe rerun the checks?

**Comment by ilyavolodin** [comment]
I restarted Travis build, maybe it was just a glitch (although doesn't feel like it). You might want to rerun all of the unittests locally, by using `mocha tests/lib` to see if our makefile is eating up errors for some reason.

**Comment by bsbeeks** [comment]
K, I'll take a look at those tests and the makefile. Stay tuned...

**Comment by nzakas** [comment]
There was a problem with the template test. That was fixed, so you should be all set if you rebase.

**Comment by bsbeeks** [comment]
@nzakas I rebased and it looks like there is still an issue. It doesn't look like a lint error, so perhaps a coverage one. How would I check coverage?

**Comment by btmills** [comment]
I pulled this down and was able to reproduce it locally. Trying to narrow down the cause right now. It looks like something's crashing and suppressing the error. Typically coverage is very vocal about a failure.

**Comment by bsbeeks** [comment]
Thanks @btmills. Let me know what you find. I'll continue to investigate as well.

**Comment by btmills** [comment]
@bsbeeks Check out the exception when you run `node_modules/.bin/mocha tests/lib/rules/max-statements` - that's the issue.

**Comment by IanVS** [comment]
@btmills any idea why the error wasn't being displayed during `npm test`?

**Comment by btmills** [comment]
@IanVS yep, see #5381!

**Comment by bsbeeks** [comment]
The test it's failing on is one where the code isn't placed into a `code` property of an object:
```
"function foo() { var a; var b; var c; var x; var y; var z; bar(); baz(); qux(); quxx(); }",
```
vs
```
{"code": "function foo() { var a; var b; var c; var x; var y; var z; bar(); baz(); qux(); quxx(); }"}
```
What use case is the this covering? I can fix the test by moving it inside an object like above, or I can update the rules to account for `context.options[0]` being `undefined`. Not sure which approach is best.

**Comment by btmills** [comment]
@bsbeeks that test covers the valid half of the default for that rule - right now, if someone enables `max-statements` without any options (i.e. `"max-statements": 2`), it defaults to 10. We only have a valid test for that, but we're missing an invalid test with 11 statements and no explicit options, which is why that wasn't obvious. In fact, it looks like `complexity`, `max-depth`, and `max-statements` all lack invalid tests with no options that verify the default behavior. So to answer your question, `"<code>"` and `{ code: "code" }` are equivalent, but the significant difference is that that test is the only one without explicit `options`. We definitely need to handle the case where `context.options[0]` is `undefined`.

A two-step approach seems to me like the best way to tackle this:

1. Submit a separate PR that adds invalid tests with no options to `complexity`, `max-depth`, and `max-statements`. Since it only adds tests without changing the implementation, we can be sure it's a non-breaking change.
2. Once (1) is merged, rebase this PR on top of that, and make sure all of the tests, both old and new, pass. The tests from (1) make sure this PR doesn't change any of the existing default behavior.

**Comment by bsbeeks** [comment]
@btmills thanks for the detailed explanation...it makes sense now! I should be able to take another stab at this later today.

**Comment by bsbeeks** [comment]
Created a PR for adding additional test for those rules (https://github.com/eslint/eslint/pull/5386). Once that is merged I'll rebase and update this PR.

File: docs/rules/max-depth.md [hunk 1]
```diff
@@ -24,6 +24,10 @@ The default depth above which this rule will warn is `4`.  You can configure the
 
 ```json
 "max-depth": [2, 10]
+
+// or you can use an object property
+
+"max-depth": [2 {"maximum": 10}]
```
Comment (jfmengels) [inline docs/rules/max-depth.md]: There is a missing comma here

Comment (bsbeeks) [inline docs/rules/max-depth.md]: Good catch, I'll get that fixed.

File: lib/rules/complexity.js [hunk 2]
```diff
@@ -11,8 +11,15 @@
...
+    var option = context.options[0],
+        THRESHOLD = 20;
 
-    var THRESHOLD = (typeof context.options[0] !== "undefined") ? context.options[0] : 20;
+    if (option.hasOwnProperty("maximum") && typeof option.maximum === "number") {
+        THRESHOLD = option.maximum;
+    }
+    if (typeof option === "number") {
+        THRESHOLD = option;
+    }
```
Comment (jfmengels) [inline lib/rules/complexity.js]: I see this exact behavior copied over multiple files. Might be useful to refactor it and put it in some common place?

Comment (bsbeeks) [inline lib/rules/complexity.js]: It appears that the pattern established in the project is to keep the logic for a rule contained within the rule file. But, I don't have enough experience with the project yet, so I'll have to refer to one of the maintainers on this.

**Comment by bsbeeks** [comment]
@btmills rebased, updated rules, and its passing CI. This should be good to go now.

File: lib/rules/complexity.js (hunk 2 above)
Comment (btmills) [inline lib/rules/complexity.js]: We don't have a good convention for sharing code between rules right now. We discussed it back in #848, but that didn't really go anywhere.

**Comment by btmills** [comment]
LGTM, thanks @bsbeeks!

**Pull request merged**